lsp/src/Language/LSP/Server/Processing.hs

```diff
@@ -449,31 +451,32 @@ handle' ::
   TClientMessage meth ->
   m ()
 handle' logger mAction m msg = do
-  maybe (return ()) (\f -> f msg) mAction
+  shutdown <- isShuttingDown
+  let allowedMethod m = case (splitClientMethod m, m) of
```

Comment:
since this is no longer local to where it's used, let's explain
```suggestion
  -- These are the methods that we are allowed to process during shutdown.
  -- The reason that we include 'shutdown' itself here is because we change 
  -- state to "shutting down" just _before_ we process the message itself!
  let allowedMethod m = case (splitClientMethod m, m) of
```